File: sdk/javascript/src/symbol/KeyPair.js
import { PrivateKey, PublicKey, Signature } from '../CryptoTypes.js';
import ed25519 from '../impl/ed25519.js';
import { deepCompare } from '../utils/arrayHelpers.js';

const HASH_MODE = 'Sha2_512';

/**
 * Represents an ED25519 private and public key.
 */
export class KeyPair {
	/**
	 * Creates a key pair from a private key.
	 * @param {PrivateKey} privateKey Private key.
	 */
	constructor(privateKey) {
		/**
		 * @private
		 */
		this._privateKey = privateKey;

		/**
		 * @private
		 */
		this._keyPair = ed25519.keyPairFromSeed(HASH_MODE, this._privateKey.bytes);
	}

	/**
	 * Gets the public key.
	 * @returns {PublicKey} Public key.
	 */
	get publicKey() {
		return new PublicKey(this._keyPair.publicKey);
	}

	/**
	 * Gets the private key.
	 * @returns {PrivateKey} Private key.
	 */
	get privateKey() {
		return new PrivateKey(this._privateKey.bytes);
	}

	/**
	 * Signs a message with the private key.
	 * @param {Uint8Array} message Message to sign.
	 * @returns {Signature} Message signature.
	 */
	sign(message) {
		return new Signature(ed25519.sign(HASH_MODE, message, this._keyPair.privateKey));
	}
}

/**
 * Verifies signatures signed by a single key pair.
 */
export class Verifier {
	/**
	 * Creates a verifier from a public key.
	 * @param {PublicKey} publicKey Public key.
	 */
	constructor(publicKey) {
		if (0 === deepCompare(new Uint8Array(PublicKey.SIZE), publicKey.bytes))
			throw new Error('public key cannot be zero');

		/**
		 * Public key used for signature verification.
		 * @type PublicKey
		 */
		this.publicKey = publicKey;
	}

	/**
	 * Verifies a message signature.
	 * @param {Uint8Array} message Message to verify.
	 * @param {Signature} signature Signature to verify.
	 * @returns {boolean} true if the message signature verifies.
	 */
	verify(message, signature) {
		return ed25519.verify(HASH_MODE, message, signature.bytes, this.publicKey.bytes);
	}
}

File: sdk/java/src/main/java/org/symbol/sdk/symbol/KeyPair.java
package org.symbol.sdk.symbol;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.impl.Ed25519;
import org.symbol.sdk.impl.Tweetnacl;

/**
 * Represents an Ed25519 private and public key for the Symbol network; uses the SHA-512 hash mode.
 */
public final class KeyPair implements org.symbol.sdk.KeyPair {
	private static final Tweetnacl.HashMode HASH_MODE = Tweetnacl.HashMode.SHA2_512;

	private final CryptoTypes.PrivateKey privateKey;
	private final Tweetnacl.KeyPair keyPair;

	/**
	 * Creates a key pair from a private key.
	 *
	 * @param privateKey Private key.
	 */
	public KeyPair(final CryptoTypes.PrivateKey privateKey) {
		this.privateKey = new CryptoTypes.PrivateKey(privateKey.bytes());
		this.keyPair = Ed25519.keyPairFromSeed(HASH_MODE, this.privateKey.bytes());
	}

	/**
	 * Gets the public key.
	 *
	 * @return Public key.
	 */
	public CryptoTypes.PublicKey getPublicKey() {
		return new CryptoTypes.PublicKey(keyPair.publicKey);
	}

	/**
	 * Gets the private key.
	 *
	 * @return Private key.
	 */
	public CryptoTypes.PrivateKey getPrivateKey() {
		return new CryptoTypes.PrivateKey(privateKey.bytes());
	}

	/**
	 * Signs a message with the private key.
	 *
	 * @param message Message to sign.
	 * @return Message signature.
	 */
	public CryptoTypes.Signature sign(final byte[] message) {
		return new CryptoTypes.Signature(Ed25519.sign(HASH_MODE, message, keyPair));
	}
}
